### PR TITLE
Deprecate exclusivePrefixSum, accumulate, adjacentDifference, iota and provide KokkosExt ones

### PR DESCRIPTION
**Pull request by aprokop**
`ArborX::exclusivePrefixSum` -> `ArborX::Details::KokkosExt::exclusive_scan`
`ArborX::accumulate` -> `ArborX::Details::KokkosExt::reduce`
`ArborX::adjacentDifference` -> `ArborX::Details::KokkosExt::adjacent_difference`
`ArborX::iota` -> `ArborX::Details::KokkosExt::iota`

Things I'm not sure about:
- Should be do `KokkosExt::StdAlgorithms` namespace?
- Should `iota` be in `StdAlgorithms`?

The goal of this and #998 is that `ArborX_DetailsUtils.hpp` is pretty clean post removal of deprecations, and nothing in `ArborX::` namespace.

### CODE REVIEW

File: test/tstDetailsUtils.cpp [hunk 1]
```diff
@@ -52,8 +53,11 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(iota, DeviceType, ARBORX_DEVICE_TYPES)
 
 BOOST_AUTO_TEST_CASE_TEMPLATE(prefix_sum, DeviceType, ARBORX_DEVICE_TYPES)
 {
+  namespace KokkosExt = ArborX::Details::KokkosExt;
```
Comment (dalg24) [inline test/tstDetailsUtils.cpp:59]:
```C++
using ArborX::Details::KokkosExt::exclusive_scan;

**Review by dalg24 — COMMENTED** [review]
Why did you not introduce the initial value trailing argument with `exclusive_scan` if the intent is to align with standard parallel algorithms?

**Comment by aprokop** [comment]
> Why did you not introduce the initial value trailing argument with `exclusive_scan` if the intent is to align with standard parallel algorithms?

Right now, the only thing I've done is moving things around. I haven't done the alignment with the standard algorithms in this PR.

File: src/details/ArborX_DetailsDistributedTreeImpl.hpp
[the remark this answers is not in the document]
Comment (aprokop) [reply src/details/ArborX_DetailsDistributedTreeImpl.hpp:1]: I'll try to add it. I do want to notice that without automatic reinforcement, we are not keeping up with the concept. Our header inclusions is often not consistent with iwyu.

File: test/tstDetailsUtils.cpp (hunk 1 above)
Comment (aprokop) [reply test/tstDetailsUtils.cpp:59]: I thought about it, but decided against it. I want to easily keep track of where `KokkosExt` and what of it is being used.

File: src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp [hunk 2]
```diff
@@ -106,6 +106,34 @@ reduce(ExecutionSpace &&space, ViewType const &v,
...
+  static_assert(SrcViewType::rank == 1 && DstViewType::rank == 1,
+                "adjacentDifference operates on rank-1 views");
+  static_assert(std::is_same<typename DstViewType::value_type,
+                             typename DstViewType::non_const_value_type>::value,
+                "adjacentDifference requires non-const destination value type");
+  static_assert(std::is_same<typename SrcViewType::non_const_value_type,
+                             typename DstViewType::value_type>::value,
+                "adjacentDifference requires same value type for source and "
+                "destination");
+  // QUESTION Should we assert anything about the memory spaces?
```
Comment (masterleinad) [inline src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp:122]: What about this?

File: src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp [hunk 3]
```diff
@@ -80,6 +80,32 @@ exclusive_scan(ExecutionSpace &&space, Kokkos::View<T, P...> const &v)
   exclusive_scan(std::forward<ExecutionSpace>(space), v, v);
 }
 
+template <typename ExecutionSpace, typename ViewType>
+typename ViewType::non_const_value_type
+reduce(ExecutionSpace &&space, ViewType const &v,
+       typename ViewType::non_const_value_type init)
+{
+  static_assert(ViewType::rank == 1, "accumulate requires a View of rank 1");
```
Comment (masterleinad) [inline src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp:88]: Check memory space accessibility as well?

File: src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp [hunk 4]
```diff
@@ -134,6 +134,27 @@ void adjacent_difference(ExecutionSpace &&space, SrcViewType const &src,
       });
 }
 
+template <typename ExecutionSpace, typename T, typename... P>
+void iota(ExecutionSpace &&space, Kokkos::View<T, P...> const &v,
+          typename Kokkos::ViewTraits<T, P...>::value_type value = 0)
+{
+  using ValueType = typename Kokkos::ViewTraits<T, P...>::value_type;
+  static_assert(unsigned(Kokkos::ViewTraits<T, P...>::rank) == unsigned(1),
+                "iota requires a View of rank 1");
```
Comment (masterleinad) [inline src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp:143]: Check memory space accessibility?

**Comment by masterleinad** [comment]
> Things I'm not sure about:
>  - Should be do KokkosExt::StdAlgorithms namespace?
>  - Should iota be in StdAlgorithms?

I don't think either of these are necessary (but it's easy to change since we don't care about backward-compatibility in that namespace).

File: src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+#include <ArborX_Exception.hpp>
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX::Details::KokkosExt
+{
+
+// NOTE: This functor is used in exclusivePrefixSum( src, dst ).  We were
+// getting a compile error on CUDA when using a KOKKOS_LAMBDA.
```
[the remark this answers is not in the document]
Comment (aprokop) [reply src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp]: I'll try.

File: test/tstUnionFind.cpp [hunk 6]
```diff
@@ -10,6 +10,7 @@
  ****************************************************************************/
 #include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
 #include "ArborX_EnableViewComparison.hpp"
+#include <ArborX_DetailsKokkosExtStdAlgorithms.hpp>
 #include <ArborX_DetailsUnionFind.hpp>
 #include <ArborX_DetailsUtils.hpp> // iota
```
Comment (dalg24) [inline test/tstUnionFind.cpp:15]: Update that comment and drop the header include if that is not needed any more

Comment (aprokop) [reply test/tstUnionFind.cpp]: Done.

File: test/tstDetailsUtils.cpp [hunk 7]
```diff
@@ -152,73 +76,20 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(minmax, DeviceType, ARBORX_DEVICE_TYPES)
...
 BOOST_AUTO_TEST_CASE_TEMPLATE(min_and_max, DeviceType, ARBORX_DEVICE_TYPES)
 {
   using ExecutionSpace = typename DeviceType::execution_space;
   ExecutionSpace space{};
 
+  namespace KokkosExt = ArborX::Details::KokkosExt;
+
   Kokkos::View<int[4], DeviceType> v("v");
-  ArborX::iota(space, v);
+  KokkosExt::iota(space, v);
```
Comment (dalg24) [inline test/tstDetailsUtils.cpp:87]: Conflicts with #998?

File: test/tstDetailsKokkosExtStdAlgorithms.cpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+BOOST_AUTO_TEST_CASE_TEMPLATE(reduce, DeviceType, ARBORX_DEVICE_TYPES)
+{
+  using ExecutionSpace = typename DeviceType::execution_space;
+  ExecutionSpace space{};
+
+  namespace KokkosExt = ArborX::Details::KokkosExt;
+
+  Kokkos::View<int[6], DeviceType> v("v");
+  Kokkos::deep_copy(v, 5);
+  BOOST_TEST(KokkosExt::reduce(space, v, 3) == 33);
```
Comment (dalg24) [inline test/tstDetailsKokkosExtStdAlgorithms.cpp:123]: add a using-declaration and do unqualified call here and below
```C++
  using ArborX::Details::KokkosExt::reduce;

File: test/tstDetailsKokkosExtStdAlgorithms.cpp [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+  KokkosExt::iota(space, w);
+  std::vector<int> w_ref = {0, 1, 2};
+  auto w_host = Kokkos::create_mirror_view(w);
+  Kokkos::deep_copy(w_host, w);
+  BOOST_TEST(w_ref == w_host, tt::per_element());
+}
+
+BOOST_AUTO_TEST_CASE_TEMPLATE(exclusive_scan, DeviceType, ARBORX_DEVICE_TYPES)
+{
+  namespace KokkosExt = ArborX::Details::KokkosExt;
```
Comment (dalg24) [inline test/tstDetailsKokkosExtStdAlgorithms.cpp:56]: Prefet
```suggestion
  using ArborX::Details::KokkosExt::exclusive_scan;
```

File: src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+          dst(i) = src(i);
+      });
+}
+
+template <typename ExecutionSpace, typename ViewType>
+void iota(ExecutionSpace const &space, ViewType const &v,
+          typename ViewType::value_type value = 0)
+{
+  static_assert(Kokkos::is_execution_space<ExecutionSpace>::value);
+  static_assert(Kokkos::is_view<ViewType>::value);
```
Comment (dalg24) [inline src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp:135]: assert that the view is accessible

File: src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+  init += tmp;
+  return init;
+}
+
+template <typename ExecutionSpace, typename SrcView, typename DstView>
+void adjacent_difference(ExecutionSpace const &space, SrcView const &src,
+                         DstView const &dst)
+{
+  static_assert(
+      Kokkos::is_execution_space<std::decay_t<ExecutionSpace>>::value);
```
Comment (dalg24) [inline src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp:97]: don't need to decay
same comment on all occurrences in this file

File: src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX::Details::KokkosExt
+{
+
+template <typename ExecutionSpace, typename SrcView, typename DstView,
+          typename InitValueType = typename DstView::value_type>
+void exclusive_scan(ExecutionSpace const &space, SrcView const &src,
+                    DstView const &dst, InitValueType init = 0)
```
Comment (dalg24) [inline src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp:26]: There is no default value in `std::exclusive_scan `

File: test/tstDetailsUtils.cpp (hunk 7 above)
Comment (aprokop) [reply test/tstDetailsUtils.cpp:87]: Yes, I will rebase #998 post-merge.

File: test/tstDetailsKokkosExtStdAlgorithms.cpp (hunk 8 above)
Comment (aprokop) [reply test/tstDetailsKokkosExtStdAlgorithms.cpp:123]: I prefer the way I did it.

File: src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp (hunk 12 above)
Comment (aprokop) [reply src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp:26]: True. It would be annoying to write 0 all the time, though.

Comment (dalg24) [reply src/kokkos_ext/ArborX_DetailsKokkosExtStdAlgorithms.hpp:26]: If we go for alignment with the standard parallel algorithms with repeating the view name for in-place scans then I don't see how specifying the init value is a big deal.

File: test/tstDetailsCrsGraphWrapperImpl.cpp [hunk 13]
```diff
@@ -11,7 +11,9 @@
 
 #include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
 #include "ArborX_EnableViewComparison.hpp"
-#include <ArborX_DetailsCrsGraphWrapperImpl.hpp>
+#include <ArborX_DetailsCrsGraphWrapperImpl.hpp> // exclusive_scan
```
Comment (dalg24) [inline test/tstDetailsCrsGraphWrapperImpl.cpp:14]: Comment is on the wrong line

File: test/tstDetailsKokkosExtStdAlgorithms.cpp [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+  Kokkos::View<int *, DeviceType> z("z", m);
+  BOOST_CHECK_THROW(KokkosExt::exclusive_scan(space, x, z, 0),
+                    ArborX::SearchException);
+  Kokkos::View<double[3], DeviceType> v("v");
+  auto v_host = Kokkos::create_mirror_view(v);
+  v_host(0) = 1.;
+  v_host(1) = 1.;
+  v_host(2) = 0.;
+  Kokkos::deep_copy(v, v_host);
+  KokkosExt::exclusive_scan(space, v, v, 5.);
```
Comment (dalg24) [inline test/tstDetailsKokkosExtStdAlgorithms.cpp:95]: Good but maybe add a comment that this is testing non-zero init value.

**Review by dalg24 — APPROVED** [review]
I would really like you to switch to using declaration in the unit test but I am not blocking
